pkg/sql/catalog/tabledesc/table_desc_builder.go

```diff
@@ -772,3 +789,203 @@ func maybeAddConstraintIDs(desc *descpb.TableDescriptor) (hasChanged bool) {
+		if err != nil {
+			return hasUpgraded, err
+		}
+	} else if tableDesc.IsView() {
+		hasUpgraded, err = maybeUpgradeSequenceReferenceForView(descLookupFn, tableDesc)
+		if err != nil {
+			return hasUpgraded, err
+		}
+	}
+
+	return hasUpgraded, err
+}
+
+// maybeUpgradeSequenceReferenceForTable upgrades all by-name sequence references
+// in `tableDesc` to by-ID.
+func maybeUpgradeSequenceReferenceForTable(
+	descLookupFn func(id descpb.ID) catalog.Descriptor, tableDesc *descpb.TableDescriptor,
+) (hasUpgraded bool, err error) {
+	if !tableDesc.IsTable() {
+		return hasUpgraded, nil
+	}
+
+	for _, col := range tableDesc.Columns {
+		// Find sequence names for all sequences used in this column.
+		usedSequenceNames, err := resolveTableNamesForIDs(descLookupFn, col.UsesSequenceIds)
+		if err != nil {
+			return hasUpgraded, err
+		}
+
+		// Upgrade sequence reference in DEFAULT expression, if any.
+		if col.HasDefault() {
+			hasUpgradedInDefault, err := seqexpr.UpgradeSequenceReferenceInExpr(col.DefaultExpr, usedSequenceNames)
+			if err != nil {
+				return hasUpgraded, err
+			}
+			hasUpgraded = hasUpgraded || hasUpgradedInDefault
+		}
+
+		// Upgrade sequence reference in ON UPDATE expression, if any.
+		if col.HasOnUpdate() {
+			hasUpgradedInOnUpdate, err := seqexpr.UpgradeSequenceReferenceInExpr(col.OnUpdateExpr, usedSequenceNames)
+			if err != nil {
+				return hasUpgraded, err
+			}
+			hasUpgraded = hasUpgraded || hasUpgradedInOnUpdate
+		}
+	}
+
+	return hasUpgraded, nil
+}
+
+// maybeUpgradeSequenceReferenceForView similarily upgrades all by-name sequence references
+// in `viewDesc` to by-ID.
+func maybeUpgradeSequenceReferenceForView(
+	descLookupFn func(id descpb.ID) catalog.Descriptor, viewDesc *descpb.TableDescriptor,
+) (hasUpgraded bool, err error) {
+	if !viewDesc.IsView() {
+		return hasUpgraded, err
+	}
+
+	// Find sequence names for all those used sequences.
+	allUsedSeqNames, err := resolveTableNamesForIDs(descLookupFn, viewDesc.DependsOn)
+	if err != nil {
+		return hasUpgraded, err
+	}
+
+	// A function that looks at an expression and replace any by-name sequence reference with
+	// by-ID reference. It, of course, also append replaced sequence IDs to `upgradedSeqIDs`.
+	replaceSeqFunc := func(expr tree.Expr) (recurse bool, newExpr tree.Expr, err error) {
+		newExprStr := expr.String()
+		hasUpgradedInExpr, err := seqexpr.UpgradeSequenceReferenceInExpr(&newExprStr, allUsedSeqNames)
+		if err != nil {
+			return false, expr, err
+		}
+		newExpr, err = parser.ParseExpr(newExprStr)
+		if err != nil {
+			return false, expr, err
+		}
+
+		hasUpgraded = hasUpgraded || hasUpgradedInExpr
+		return false, newExpr, err
+	}
+
+	stmt, err := parser.ParseOne(viewDesc.GetViewQuery())
+	if err != nil {
+		return hasUpgraded, err
+	}
+
+	newStmt, err := tree.SimpleStmtVisit(stmt.AST, replaceSeqFunc)
+	if err != nil {
+		return hasUpgraded, err
+	}
+
+	viewDesc.ViewQuery = newStmt.String()
```

Comment:
I'd like @ajwerner 's opinion on whether it's safe to modify the view query like this or not, just to make sure. Personally I don't see why not.